File: 532.k-diff-pairs-in-an-array.ts
/*
 * @lc app=leetcode id=532 lang=typescript
 *
 * [532] K-diff Pairs in an Array
 */

// @lc code=start
function findPairs(nums: number[], k: number): number {
  const seen = new Set<number>()
  const uniquePairs = new Set<number>()

  for (const num of nums) {
    if (seen.has(num - k)) {
      uniquePairs.add(num - k)
    }

    if (seen.has(num + k)) {
      uniquePairs.add(num)
    }
    seen.add(num)
  }

  return uniquePairs.size
}
// @lc code=end

File: go/532.k-diff-pairs-in-an-array.go
func findPairs(nums []int, k int) int {
    pairs := make(map[int]int)
    result := 0
    for _, v := range nums {
        pairs[v] += 1
    }
    for key, v := range pairs {
        if k == 0  {
            if v >= 2{
                result += 1
            }
        } else {
            _, ok := pairs[key + k]
            if ok {
                result += 1
            }
        }
    }
    return result
}